linkerd/lock/src/lib.rs

```diff
@@ -0,0 +1,363 @@
+//! A middleware for sharing an inner service via mutual exclusion.
+
+#![deny(warnings, rust_2018_idioms)]
+
+use futures::{future, Async, Future, Poll};
+use linkerd2_error::Error;
+use tokio::sync::lock;
+use tracing::trace;
+
+#[derive(Clone, Debug)]
+pub struct Layer<E = Poisoned> {
+    _marker: std::marker::PhantomData<E>,
+}
+
+/// Guards access to an inner service with a `tokio::sync::lock::Lock`.
+///
+/// As the service is polled to readiness, the lock is acquired and the inner
+/// service is polled. If the sevice is cloned, the service's lock state is not
+/// retained by the clone.
+///
+/// The inner service's errors are coerced to the cloneable `C`-typed error so
+/// that the error may be returned to all clones of the lock. By default, errors
+/// are propagated through the `Poisoned` type, but they may be propagated
+/// through custom types as well.
+pub struct Lock<S, E = Poisoned> {
+    lock: lock::Lock<State<S, E>>,
+    locked: Option<lock::LockGuard<State<S, E>>>,
+}
+
+/// The lock holds either an inner srvice or, if it failed, an error salvaged
+/// from the failure.
+enum State<S, E> {
+    Service(S),
+    Error(E),
+}
+
+/// A default error type that propagates the inner service's error message to
+/// consumers.
+#[derive(Clone, Debug)]
+pub struct Poisoned(String);
+
+// === impl Layer ===
+
+impl Default for Layer {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Layer {
+    /// Sets the error type to be returned to consumers when poll_ready fails.
+    pub fn new<E>() -> Layer<E>
+    where
+        E: Clone + From<Error> + Into<Error>,
```

Comment:
I don't think that's quite right. See [the playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=763b5537dec8e54d4331d3d46940133b). Just because we can `E::from(Error)` doesn't mean we can `Error::from(E)`.

Reply:
Ah right it provides `Into<E> for Error` and we need `Into<Error> for E>`. Sorry about that!